chat_with_nerf/visual_grounder/captioner.py

```diff
@@ -0,0 +1,158 @@
+from typing import Any, Optional
+
+import torch
+from attrs import Factory, define
+from llava.conversation import SeparatorStyle, conv_templates
+from llava.model import *  # noqa: F401, F403
+from llava.model.utils import KeywordsStoppingCriteria
+from PIL import Image
+
+from chat_with_nerf.settings import Settings
+from chat_with_nerf.visual_grounder.image_ref import ImageRef
+
+
+@define
+class BaseCaptioner:
+    model: torch.nn.Module
+    """Base model for image processing."""
+    vis_processors: dict
+    """Preprocessors for visual inputs."""
+    images: list[ImageRef] = Factory(list)
+
+    def load_images(self, selected: list[ImageRef]) -> None:
+        """Loads images from the provided list of ImageRefs."""
+        # load sample image
+        for image_ref in selected:
+            raw_image = Image.open(image_ref.rgb_address).convert("RGB")
+            # raw_image.resize((596, 437))
+            image_ref.raw_image = raw_image
+        self.images = selected
+
+    def process_image(self, image_path: str) -> torch.Tensor:
+        """Processes an image and returns it as a tensor."""
+        raw_image = Image.open(image_path).convert("RGB")
+        # raw_image.resize((596, 437))
+        return self.vis_processors["eval"](raw_image).unsqueeze(0).to(self.model.device)
+
+    def caption(self):
+        """Generates captions for the images."""
+        pass
+
+
+@define
+class Blip2Captioner(BaseCaptioner):
+    positive_words: str = "computer"
+
+    def set_positive_words(self, new_positive_words):
+        self.positive_words = new_positive_words
+
+    def caption(self) -> dict[str, str]:
+        """_summary_
+
+        :return: a dictionary of image path and its corresponding caption
+        :rtype: dict[str, str]
+        """
+        # TODO: batch it.
+        # prepare the image
+        result: dict[str, str] = {}  # key: rgb_address, value: caption
+        for image_ref in self.images:
+            image = self.process_image(image_ref.raw_image)  # type: ignore
+            question = (
+                "Describe the shape and material of the"
+                + self.positive_words
+                + ", if there is one."
+            )
+            answer = self.model.generate({"image": image, "prompt": question})[0]  # type: ignore
+            result[image_ref.rgb_address] = answer
+
+        return result
+
+
+@define
+class LLaVaCaptioner(BaseCaptioner):
+    tokenizer: Optional[Any] = None
+    mm_use_im_start_end: bool = True
+    image_token_len: int = 512
+    positive_words: str = "computer"
+
+    def set_positive_words(self, new_positive_words):
+        self.positive_words = new_positive_words
```

Comment:
delete `set_positive_words()` method and make `caption()` take in `positive_words`.